redact/index.ts

```diff
@@ -0,0 +1,208 @@
+  replacement: string,
+  start: number,
+  end: number,
+): string {
+  return text.substring(0, start) + replacement + text.substring(end);
+}
+
+function noOpDetect(_tokens: string[]): Array<SensitiveInfoEntity | undefined> {
+  return [];
+}
+function noOpReplace(_input: SensitiveInfoEntity): string | undefined {
+  return undefined;
+}
+
+interface RedactedSensitiveInfoEntity {
+  original: string;
+  redacted: string;
+  start: number;
+  end: number;
+  identifiedType: string;
+}
+
+async function callRedactWasm<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
+  Replace extends ReplaceSensitiveInfoEntities<CustomEntities>,
+  CustomEntities extends string,
+>(
+  candidate: string,
+  options: RedactOptions<Detect, Replace, CustomEntities>,
+): Promise<RedactedSensitiveInfoEntity[]> {
+  let convertedDetect = noOpDetect;
+  if (typeof options.detect === "function") {
+    const detect = options.detect;
+    convertedDetect = (tokens: string[]) => {
+      return detect(tokens)
+        .filter((e) => typeof e !== "undefined")
+        .map((e) => userEntitiesToWasm(e));
+    };
+  }
+
+  let convertedReplace = noOpReplace;
+  if (typeof options.replace === "function") {
+    const replace = options.replace;
+    convertedReplace = (identifiedType: SensitiveInfoEntity) => {
+      // We need to use an `as` here because the Wasm generated types just use `string` for custom.
+      return replace(
+        wasmEntitiesToString(identifiedType) as
+          | CustomEntities
+          | ArcjetSensitiveInfoType,
+      );
+    };
+  }
+
+  const wasm = await initializeWasm(convertedDetect, convertedReplace);
+
+  if (typeof wasm !== "undefined") {
+    const skipCustomDetect = typeof options.detect !== "function";
+    const skipCustomRedact = typeof options.replace !== "function";
+
+    const config = {
+      entities: options.redact.map(userEntitiesToWasm),
+      contextWindowSize: options.contextWindowSize,
+      skipCustomDetect,
+      skipCustomRedact,
+    };
+
+    return wasm.redact(candidate, config).map((e) => {
+      return {
+        ...e,
+        identifiedType: wasmEntitiesToString(e.identifiedType),
+      };
+    });
+  } else {
+    throw new Error(
+      "redact failed to run because Wasm is not supported in this environment",
+    );
+  }
+}
+
+type Unredact = (input: string) => string;
+
+export async function redact<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
+  Replace extends ReplaceSensitiveInfoEntities<CustomEntities>,
+  CustomEntities extends string,
+>(
+  candidate: string,
+  options: RedactOptions<Detect, Replace, CustomEntities>,
+): Promise<[string, Unredact]> {
+  const redactions = await callRedactWasm(candidate, options);
+
+  // Need to apply the redactions in reverse order so that the offsets aren't changed
+  // when we redact with strings that are longer/shorter than the original.
+  redactions.reverse();
+  for (const redaction of redactions) {
```

Comment:
```suggestion

  for (const redaction of redactions) {
```
Nit: having a space here seems like I'd look at the comment and reverse line above.